File: app/src/main/java/huce/fit/appreadstories/adapters/ChapterAdapter.java
package huce.fit.appreadstories.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import huce.fit.appreadstories.R;
import huce.fit.appreadstories.interfaces.ClickListener;
import huce.fit.appreadstories.model.ChuongTruyen;

public class ChapterAdapter extends RecyclerView.Adapter<ChapterAdapter.ChapterHoder> {
    private final List<ChuongTruyen> listChapter;
    private final List<ChuongTruyen> listChapterReadOld;
    private final List<ChuongTruyen> listChapterRead = new ArrayList<>();
    private final ClickListener clickListener;
    private final Context context;
    private final int idChapterReading;

    public ChapterAdapter(Context context, List<ChuongTruyen> listChapter, List<ChuongTruyen> listChapterRead, int idChapterReading, ClickListener clickListener) {
        this.context = context;
        this.listChapter = listChapter;
        this.listChapterReadOld = listChapterRead;
        this.idChapterReading = idChapterReading;
        this.clickListener = clickListener;
    }

    @NonNull
    @Override
    public ChapterHoder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.activity_chapter_item, parent, false);
        return new ChapterHoder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ChapterHoder holder, int position) {
        int count = 0;

        holder.itemView.setOnClickListener(view -> {
            int idChapter = listChapter.get(position).getMachuong();
            clickListener.onItemClick(idChapter, view, false);
        });

        if (listChapterRead.size() == 0) {
            if (count == 0) {
                for (ChuongTruyen c : listChapterReadOld) {
                    if (c.getMachuong() != idChapterReading) {
                        listChapterRead.add(c);
                    }
                }
            } else {
                listChapterRead.clear();
            }
        }


        ChuongTruyen ct = listChapter.get(position);
        if (ct != null) {
            holder.tvNumberChapter.setText(String.format(Locale.getDefault(),"%s.", ct.getSochuong()));
            holder.tvChapter.setText(ct.getTenchuong());
            holder.tvPostDay.setText(ct.getThoigiandang());

            int idChapter = ct.getMachuong();

            if (idChapter == idChapterReading) {
                holder.tvNumberChapter.setTextColor(context.getResources().getColor(R.color.orange));
                holder.tvChapter.setTextColor(context.getResources().getColor(R.color.orange));
                holder.tvPostDay.setTextColor(context.getResources().getColor(R.color.orange));
            }

            //so sánh mã chương trong danh sách truyện đã đọc với mã chương
            if (listChapterRead.size() > 0 && idChapterReading > 0) {
                for (int i = 0; i < listChapterRead.size(); i++) {
                    int idChapterRead = listChapterRead.get(i).getMachuong();

                    if (idChapterRead == idChapter) {
                        holder.tvNumberChapter.setTextColor(context.getResources().getColor(R.color.dim_gray));
                        holder.tvChapter.setTextColor(context.getResources().getColor(R.color.dim_gray));
                        holder.tvPostDay.setTextColor(context.getResources().getColor(R.color.dim_gray));

                        listChapterRead.remove(listChapterRead.get(i));
                        count++;
                        break;
                    }
                    if (idChapterRead != idChapter && idChapterReading != idChapter) {
                        holder.tvNumberChapter.setTextColor(context.getResources().getColor(R.color.black));
                        holder.tvChapter.setTextColor(context.getResources().getColor(R.color.black));
                        holder.tvPostDay.setTextColor(context.getResources().getColor(R.color.black));
                    }
                }
            }
        }
    }

    @Override
    public int getItemCount() {
        if (listChapter != null && listChapter.size() > 0)
            return listChapter.size();
        else
            return 0;
    }

    public class ChapterHoder extends RecyclerView.ViewHolder {
        private TextView tvNumberChapter, tvChapter, tvPostDay;

        public ChapterHoder(@NonNull View itemView) {
            super(itemView);
            tvNumberChapter = itemView.findViewById(R.id.tvNumberChapter);
            tvChapter = itemView.findViewById(R.id.tvChapter);
            tvPostDay = itemView.findViewById(R.id.tvPostDay);
        }
    }
}

File: app/src/main/java/huce/fit/appreadstories/adapters/ChapterAdapter.kt
package huce.fit.appreadstories.adapters

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import huce.fit.appreadstories.R
import huce.fit.appreadstories.model.Chapter
import huce.fit.appreadstories.model.ChapterRead
import java.util.Locale

class ChapterAdapter(private var context: Context) : RecyclerView.Adapter<ChapterAdapter.ChapterHolder>() , Filterable {

    companion object{
        private const val TAG = "ChapterAdapter"
    }

    private var chapterList: MutableList<Chapter> = mutableListOf()
    private var chapterListOld: MutableList<Chapter> = mutableListOf()
    private val chapterReadMap = HashMap<Int, ChapterRead?>()
    private lateinit var clickListener: ClickListener
    private var chapterReadingId = 0

    class ChapterHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val tvChapterNumber: TextView = itemView.findViewById(R.id.tvChapterNumber)
        val tvChapterName: TextView = itemView.findViewById(R.id.tvChapterName)
        val tvPostDay: TextView= itemView.findViewById(R.id.tvPostDay)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ChapterHolder {
        val view: View = LayoutInflater.from(parent.context)
            .inflate(R.layout.activity_chapter_item, parent, false)
        return ChapterHolder(view)
    }

    override fun onBindViewHolder(holder: ChapterHolder, position: Int) {
        val chapter = chapterList[position]
        val chapterId = chapter.chapterId
        holder.tvChapterNumber.text = String.format(
            Locale.getDefault(), "%s.", chapter.chapterNumber
        )
        holder.tvChapterName.text = chapter.chapterName
        holder.tvPostDay.text = chapter.postDay
        holder.itemView.setOnClickListener { view: View ->
            clickListener.onItemClick(chapterId, view, false)
        }
        if (chapterReadMap.containsKey(chapterId)) {
            if (chapterReadingId != chapterId) {
                holder.tvChapterNumber.setTextColor(context.getColor(R.color.dim_gray))
                holder.tvChapterName.setTextColor(context.getColor(R.color.dim_gray))
                holder.tvPostDay.setTextColor(context.getColor(R.color.dim_gray))
            } else {
                holder.tvChapterNumber.setTextColor(context.getColor(R.color.orange))
                holder.tvChapterName.setTextColor(context.getColor(R.color.orange))
                holder.tvPostDay.setTextColor(context.getColor(R.color.orange))
            }
        } else {
            holder.tvChapterNumber.setTextColor(context.getColor(R.color.black))
            holder.tvChapterName.setTextColor(context.getColor(R.color.black))
            holder.tvPostDay.setTextColor(context.getColor(R.color.black))
        }
    }

    override fun getItemCount(): Int {
        return chapterList.size
    }


    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(charSequence: CharSequence): FilterResults {
                val searchStr = charSequence.toString()
                chapterList = if (searchStr.isEmpty()) {
                    chapterListOld
                } else {
                    val list: MutableList<Chapter> = mutableListOf()
                    for (chapter in chapterListOld) {
                        if (chapter.chapterNumber.contains(searchStr)) {
                            list.add(chapter)
                        }
                    }
                    list
                }
                val filterResults = FilterResults()
                filterResults.values = chapterList
                return filterResults
            }

            @SuppressLint("NotifyDataSetChanged")
            override fun publishResults(charSequence: CharSequence, filterResults: FilterResults) {
                chapterList = filterResults.values as MutableList<Chapter>
                notifyDataSetChanged()
            }
        }
    }

    fun clickListener(clickListener: ClickListener) {
        this.clickListener = clickListener
    }

    fun reverseData() {
        chapterList.reverse()
    }

    @SuppressLint("NotifyDataSetChanged")
    fun setChapterList(chapterList: MutableList<Chapter>) {
        this.chapterList = chapterList
        chapterListOld = chapterList
        notifyDataSetChanged()
    }

    @SuppressLint("NotifyDataSetChanged")
    fun setChapterReadList(chapterReadList: List<ChapterRead>, chapterReadingId: Int) {
        for (chapterRead in chapterReadList) {
            chapterReadMap[chapterRead.chapterId] = null
        }
        this.chapterReadingId = chapterReadingId
        notifyDataSetChanged()
    }
}
